diff --git a/.prettierignore b/.prettierignore
index 2080c85..87b7e86 100644
--- a/.prettierignore
+++ b/.prettierignore
@@ -1,2 +1,3 @@
 deploymentFiles/
 dlc-solidity/deploymentFiles/
+.openzeppelin/
diff --git a/README.md b/README.md
index 447c1d0..5570a31 100644
--- a/README.md
+++ b/README.md
@@ -15,7 +15,7 @@ This repo contains the solidity smart contracts for the DLC.Link infrastructure
 
 Learn more about [DLCs](https://github.com/DLC-link/dlc-solidity#What-Are-DLCs) and [dlcBTC](https://dlcbtc.com).
 
-## Overview
+# Overview
 
 A DLC is a contract on Bitcoin that enables users to move/lock Bitcoin conditionally. The possible outcomes of a DLC are predefined and stored by the DLC.Link Attestor Layer. Bitcoin locked in such a way can be represented then on any EVM chain using these smart contracts. The outcome - that is, the value that will be 'attested' to - is supplied by smart contracts too.
 
@@ -26,6 +26,12 @@ _dlcBTC_ is a token that represents Bitcoin locked in DLCs. It is minted when Bi
 Learn more about the whole architecture on the documentation site here:
 https://docs.dlc.link/architecture/tech-stack
 
+## Timelocked Upgrades
+
+Some of the contracts in this repository are upgradeable. They can be changed by a multisig of dlcBTC team and external members. The upgrade process is timelocked, meaning that the upgrade will only happen after a certain time has passed (minimum 7 days). This is to ensure that the community has time to review the changes and react if necessary.
+
+<img src="./assets/timelockArchitecture.svg" alt="architecture" width="100%" />
+
 # Dev notes
 
 ## Testing
diff --git a/contracts/DLCManager.sol b/contracts/DLCManager.sol
index 37ce286..d125b85 100644
--- a/contracts/DLCManager.sol
+++ b/contracts/DLCManager.sol
@@ -13,14 +13,14 @@ import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
 import "@openzeppelin/contracts-upgradeable/utils/cryptography/ECDSAUpgradeable.sol";
 import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 import "./DLCLinkLibrary.sol";
-import "./DLCBTC.sol";
+import "./IBTC.sol";
 
 import "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";
 
 /**
  * @author  DLC.Link 2024
  * @title   DLCManager
- * @dev     This is the contract the Attestor Layer listens to.
+ * @dev     This is the contract the Attestor Layer listens and writes to
  * @dev     It is upgradable through the OpenZeppelin proxy pattern
  * @notice  DLCManager is the main contract of the DLC.Link protocol.
  * @custom:contact eng@dlc.link
@@ -33,7 +33,7 @@ contract DLCManager is
 {
     using DLCLink for DLCLink.DLC;
     using DLCLink for DLCLink.DLCStatus;
-    using SafeERC20 for DLCBTC;
+    using SafeERC20 for IBTC;
 
     ////////////////////////////////////////////////////////////////
     //                      STATE VARIABLES                       //
@@ -56,7 +56,9 @@ contract DLCManager is
     bytes32 public tssCommitment;
     string public attestorGroupPubKey;
 
-    DLCBTC public dlcBTC; // dlcBTC contract
+    // iBTC was historically called dlcBTC.
+    // Due the nature of upgradability, we have to keep the old name.
+    IBTC public dlcBTC; // iBTC contract.
     string public btcFeeRecipient; // BTC address to send fees to
     uint256 public minimumDeposit; // in sats
     uint256 public maximumDeposit; // in sats
@@ -69,7 +71,8 @@ contract DLCManager is
     bool public porEnabled;
     AggregatorV3Interface public dlcBTCPoRFeed;
     mapping(address => mapping(bytes32 => bool)) private _seenSigners;
-    uint256[39] __gap;
+    uint256 public totalValueMinted;
+    uint256[38] __gap;
 
     ////////////////////////////////////////////////////////////////
     //                           ERRORS                           //
@@ -135,7 +138,7 @@ contract DLCManager is
         address defaultAdmin,
         address dlcAdminRole,
         uint16 threshold,
-        DLCBTC tokenContract,
+        IBTC tokenContract,
         string memory btcFeeRecipientToSet
     ) public initializer {
         __AccessControlDefaultAdminRules_init(2 days, defaultAdmin);
@@ -154,6 +157,21 @@ contract DLCManager is
         btcRedeemFeeRate = 15; // 0.15% BTC fee for now
         btcFeeRecipient = btcFeeRecipientToSet;
         porEnabled = false;
+        totalValueMinted = 0;
+    }
+
+    /**
+     * @notice Initialize total minted value tracking
+     * @dev    This function is called once after the contract is upgraded with totalValueMinted tracking
+     */
+    function initializeV2() public reinitializer(2) {
+        // Calculate initial total by iterating through existing vaults
+        uint256 total = 0;
+        for (uint256 i = 0; i < _index; i++) {
+            total += dlcs[i].valueMinted;
+        }
+
+        totalValueMinted = total;
     }
 
     /// @custom:oz-upgrades-unsafe-allow constructor
@@ -400,9 +418,6 @@ contract DLCManager is
             revert DepositTooSmall(amountToLockDiff, minimumDeposit);
         }
 
-        // We fetch the current total minted value in all vaults before we update this Vault
-        uint256 currentTotalMinted = getTotalValueMintedInVaults();
-
         dlc.fundingTxId = btcTxId;
         dlc.wdTxId = "";
         dlc.status = DLCLink.DLCStatus.FUNDED;
@@ -410,7 +425,8 @@ contract DLCManager is
         dlc.valueLocked = newValueLocked;
         dlc.valueMinted = newValueLocked;
 
-        if (_checkMint(amountToMint, currentTotalMinted)) {
+        if (_checkMint(amountToMint, totalValueMinted)) {
+            totalValueMinted = totalValueMinted + amountToMint;
             _mintTokens(dlc.creator, amountToMint);
         }
 
@@ -489,6 +505,7 @@ contract DLCManager is
         }
 
         dlc.valueMinted -= amount;
+        totalValueMinted -= amount;
         _burnTokens(dlc.creator, amount);
         emit Withdraw(uuid, amount, msg.sender);
     }
@@ -555,14 +572,6 @@ contract DLCManager is
         return vaults;
     }
 
-    function getTotalValueMintedInVaults() public view returns (uint256) {
-        uint256 totalValueMinted = 0;
-        for (uint256 i = 0; i < _index; i++) {
-            totalValueMinted += dlcs[i].valueMinted;
-        }
-        return totalValueMinted;
-    }
-
     function isWhitelisted(address account) external view returns (bool) {
         return _whitelistedAddresses[account];
     }
diff --git a/contracts/DLCBTC.sol b/contracts/IBTC.sol
similarity index 73%
rename from contracts/DLCBTC.sol
rename to contracts/IBTC.sol
index fd09984..fce48d2 100644
--- a/contracts/DLCBTC.sol
+++ b/contracts/IBTC.sol
@@ -14,13 +14,14 @@ import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 
 /**
  * @author  DLC.Link 2024
- * @title   DLCBTC
- * @notice  The DLCBTC Token represents Bitcoin locked through the DLC.Link bridge
- * @dev     Owner is the TokenManager contract
+ * @title   iBTC
+ * @notice  The iBTC Token represents Bitcoin locked in self-custody by the DLC.Link protocol.
+ * @dev     Owner is the DLCManager contract, which can mint/burn tokens
+ * @dev     Minter/Burner rights are also given to CCIP token pools
  * @custom:contact eng@dlc.link
  * @custom:website https://www.dlc.link
  */
-contract DLCBTC is
+contract IBTC is
     Initializable,
     ERC20Upgradeable,
     ERC20PermitUpgradeable,
@@ -42,9 +43,18 @@ contract DLCBTC is
     }
 
     function initialize() public initializer {
-        __ERC20_init("dlcBTC", "dlcBTC");
+        __ERC20_init("iBTC", "IBTC");
         __Ownable_init();
-        __ERC20Permit_init("dlcBTC");
+        __ERC20Permit_init("iBTC");
+    }
+
+    /**
+     * @notice Reinitializes the EIP712 domain separator
+     * @dev This function is used to reinitialize the EIP712 domain separator after a change in the name
+     * @dev Name changed from "dlcBTC" to "iBTC"
+     */
+    function reinitializeEIP712() public reinitializer(2) {
+        __EIP712_init_unchained("iBTC", "1");
     }
 
     modifier onlyOwnerOrCCIPMinter() {
@@ -63,8 +73,12 @@ contract DLCBTC is
         return 8;
     }
 
+    function name() public view virtual override returns (string memory) {
+        return "iBTC";
+    }
+
     function symbol() public view virtual override returns (string memory) {
-        return "dlcBTC";
+        return "IBTC";
     }
 
     function mint(address to, uint256 amount) external onlyOwnerOrCCIPMinter {
diff --git a/contracts/TimelockController.sol b/contracts/TimelockController.sol
new file mode 100644
index 0000000..a82a6f1
--- /dev/null
+++ b/contracts/TimelockController.sol
@@ -0,0 +1,505 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.9.0) (governance/TimelockController.sol)
+
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import "@openzeppelin/contracts/token/ERC1155/IERC1155Receiver.sol";
+
+/**
+ * @dev Contract module which acts as a timelocked controller. When set as the
+ * owner of an `Ownable` smart contract, it enforces a timelock on all
+ * `onlyOwner` maintenance operations. This gives time for users of the
+ * controlled contract to exit before a potentially dangerous maintenance
+ * operation is applied.
+ *
+ * By default, this contract is self administered, meaning administration tasks
+ * have to go through the timelock process. The proposer (resp executor) role
+ * is in charge of proposing (resp executing) operations. A common use case is
+ * to position this {TimelockController} as the owner of a smart contract, with
+ * a multisig or a DAO as the sole proposer.
+ *
+ * _Available since v3.3._
+ */
+contract TimelockController is
+    AccessControl,
+    IERC721Receiver,
+    IERC1155Receiver
+{
+    bytes32 public constant TIMELOCK_ADMIN_ROLE =
+        keccak256("TIMELOCK_ADMIN_ROLE");
+    bytes32 public constant PROPOSER_ROLE = keccak256("PROPOSER_ROLE");
+    bytes32 public constant EXECUTOR_ROLE = keccak256("EXECUTOR_ROLE");
+    bytes32 public constant CANCELLER_ROLE = keccak256("CANCELLER_ROLE");
+    uint256 internal constant _DONE_TIMESTAMP = uint256(1);
+
+    mapping(bytes32 => uint256) private _timestamps;
+    uint256 private _minDelay;
+
+    /**
+     * @dev Emitted when a call is scheduled as part of operation `id`.
+     */
+    event CallScheduled(
+        bytes32 indexed id,
+        uint256 indexed index,
+        address target,
+        uint256 value,
+        bytes data,
+        bytes32 predecessor,
+        uint256 delay
+    );
+
+    /**
+     * @dev Emitted when a call is performed as part of operation `id`.
+     */
+    event CallExecuted(
+        bytes32 indexed id,
+        uint256 indexed index,
+        address target,
+        uint256 value,
+        bytes data
+    );
+
+    /**
+     * @dev Emitted when new proposal is scheduled with non-zero salt.
+     */
+    event CallSalt(bytes32 indexed id, bytes32 salt);
+
+    /**
+     * @dev Emitted when operation `id` is cancelled.
+     */
+    event Cancelled(bytes32 indexed id);
+
+    /**
+     * @dev Emitted when the minimum delay for future operations is modified.
+     */
+    event MinDelayChange(uint256 oldDuration, uint256 newDuration);
+
+    /**
+     * @dev Initializes the contract with the following parameters:
+     *
+     * - `minDelay`: initial minimum delay for operations
+     * - `proposers`: accounts to be granted proposer and canceller roles
+     * - `executors`: accounts to be granted executor role
+     * - `admin`: optional account to be granted admin role; disable with zero address
+     *
+     * IMPORTANT: The optional admin can aid with initial configuration of roles after deployment
+     * without being subject to delay, but this role should be subsequently renounced in favor of
+     * administration through timelocked proposals. Previous versions of this contract would assign
+     * this admin to the deployer automatically and should be renounced as well.
+     */
+    constructor(
+        uint256 minDelay,
+        address[] memory proposers,
+        address[] memory executors,
+        address admin
+    ) {
+        _setRoleAdmin(TIMELOCK_ADMIN_ROLE, TIMELOCK_ADMIN_ROLE);
+        _setRoleAdmin(PROPOSER_ROLE, TIMELOCK_ADMIN_ROLE);
+        _setRoleAdmin(EXECUTOR_ROLE, TIMELOCK_ADMIN_ROLE);
+        _setRoleAdmin(CANCELLER_ROLE, TIMELOCK_ADMIN_ROLE);
+
+        // self administration
+        _setupRole(TIMELOCK_ADMIN_ROLE, address(this));
+
+        // optional admin
+        if (admin != address(0)) {
+            _setupRole(TIMELOCK_ADMIN_ROLE, admin);
+        }
+
+        // register proposers and cancellers
+        for (uint256 i = 0; i < proposers.length; ++i) {
+            _setupRole(PROPOSER_ROLE, proposers[i]);
+            _setupRole(CANCELLER_ROLE, proposers[i]);
+        }
+
+        // register executors
+        for (uint256 i = 0; i < executors.length; ++i) {
+            _setupRole(EXECUTOR_ROLE, executors[i]);
+        }
+
+        _minDelay = minDelay;
+        emit MinDelayChange(0, minDelay);
+    }
+
+    /**
+     * @dev Modifier to make a function callable only by a certain role. In
+     * addition to checking the sender's role, `address(0)` 's role is also
+     * considered. Granting a role to `address(0)` is equivalent to enabling
+     * this role for everyone.
+     */
+    modifier onlyRoleOrOpenRole(bytes32 role) {
+        if (!hasRole(role, address(0))) {
+            _checkRole(role, _msgSender());
+        }
+        _;
+    }
+
+    /**
+     * @dev Contract might receive/hold ETH as part of the maintenance process.
+     */
+    receive() external payable {}
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(
+        bytes4 interfaceId
+    ) public view virtual override(IERC165, AccessControl) returns (bool) {
+        return
+            interfaceId == type(IERC1155Receiver).interfaceId ||
+            super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev Returns whether an id correspond to a registered operation. This
+     * includes both Pending, Ready and Done operations.
+     */
+    function isOperation(bytes32 id) public view virtual returns (bool) {
+        return getTimestamp(id) > 0;
+    }
+
+    /**
+     * @dev Returns whether an operation is pending or not. Note that a "pending" operation may also be "ready".
+     */
+    function isOperationPending(bytes32 id) public view virtual returns (bool) {
+        return getTimestamp(id) > _DONE_TIMESTAMP;
+    }
+
+    /**
+     * @dev Returns whether an operation is ready for execution. Note that a "ready" operation is also "pending".
+     */
+    function isOperationReady(bytes32 id) public view virtual returns (bool) {
+        uint256 timestamp = getTimestamp(id);
+        return timestamp > _DONE_TIMESTAMP && timestamp <= block.timestamp;
+    }
+
+    /**
+     * @dev Returns whether an operation is done or not.
+     */
+    function isOperationDone(bytes32 id) public view virtual returns (bool) {
+        return getTimestamp(id) == _DONE_TIMESTAMP;
+    }
+
+    /**
+     * @dev Returns the timestamp at which an operation becomes ready (0 for
+     * unset operations, 1 for done operations).
+     */
+    function getTimestamp(bytes32 id) public view virtual returns (uint256) {
+        return _timestamps[id];
+    }
+
+    /**
+     * @dev Returns the minimum delay for an operation to become valid.
+     *
+     * This value can be changed by executing an operation that calls `updateDelay`.
+     */
+    function getMinDelay() public view virtual returns (uint256) {
+        return _minDelay;
+    }
+
+    /**
+     * @dev Returns the identifier of an operation containing a single
+     * transaction.
+     */
+    function hashOperation(
+        address target,
+        uint256 value,
+        bytes calldata data,
+        bytes32 predecessor,
+        bytes32 salt
+    ) public pure virtual returns (bytes32) {
+        return keccak256(abi.encode(target, value, data, predecessor, salt));
+    }
+
+    /**
+     * @dev Returns the identifier of an operation containing a batch of
+     * transactions.
+     */
+    function hashOperationBatch(
+        address[] calldata targets,
+        uint256[] calldata values,
+        bytes[] calldata payloads,
+        bytes32 predecessor,
+        bytes32 salt
+    ) public pure virtual returns (bytes32) {
+        return
+            keccak256(abi.encode(targets, values, payloads, predecessor, salt));
+    }
+
+    /**
+     * @dev Schedule an operation containing a single transaction.
+     *
+     * Emits {CallSalt} if salt is nonzero, and {CallScheduled}.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'proposer' role.
+     */
+    function schedule(
+        address target,
+        uint256 value,
+        bytes calldata data,
+        bytes32 predecessor,
+        bytes32 salt,
+        uint256 delay
+    ) public virtual onlyRole(PROPOSER_ROLE) {
+        bytes32 id = hashOperation(target, value, data, predecessor, salt);
+        _schedule(id, delay);
+        emit CallScheduled(id, 0, target, value, data, predecessor, delay);
+        if (salt != bytes32(0)) {
+            emit CallSalt(id, salt);
+        }
+    }
+
+    /**
+     * @dev Schedule an operation containing a batch of transactions.
+     *
+     * Emits {CallSalt} if salt is nonzero, and one {CallScheduled} event per transaction in the batch.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'proposer' role.
+     */
+    function scheduleBatch(
+        address[] calldata targets,
+        uint256[] calldata values,
+        bytes[] calldata payloads,
+        bytes32 predecessor,
+        bytes32 salt,
+        uint256 delay
+    ) public virtual onlyRole(PROPOSER_ROLE) {
+        require(
+            targets.length == values.length,
+            "TimelockController: length mismatch"
+        );
+        require(
+            targets.length == payloads.length,
+            "TimelockController: length mismatch"
+        );
+
+        bytes32 id = hashOperationBatch(
+            targets,
+            values,
+            payloads,
+            predecessor,
+            salt
+        );
+        _schedule(id, delay);
+        for (uint256 i = 0; i < targets.length; ++i) {
+            emit CallScheduled(
+                id,
+                i,
+                targets[i],
+                values[i],
+                payloads[i],
+                predecessor,
+                delay
+            );
+        }
+        if (salt != bytes32(0)) {
+            emit CallSalt(id, salt);
+        }
+    }
+
+    /**
+     * @dev Schedule an operation that is to become valid after a given delay.
+     */
+    function _schedule(bytes32 id, uint256 delay) private {
+        require(
+            !isOperation(id),
+            "TimelockController: operation already scheduled"
+        );
+        require(
+            delay >= getMinDelay(),
+            "TimelockController: insufficient delay"
+        );
+        _timestamps[id] = block.timestamp + delay;
+    }
+
+    /**
+     * @dev Cancel an operation.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'canceller' role.
+     */
+    function cancel(bytes32 id) public virtual onlyRole(CANCELLER_ROLE) {
+        require(
+            isOperationPending(id),
+            "TimelockController: operation cannot be cancelled"
+        );
+        delete _timestamps[id];
+
+        emit Cancelled(id);
+    }
+
+    /**
+     * @dev Execute an (ready) operation containing a single transaction.
+     *
+     * Emits a {CallExecuted} event.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'executor' role.
+     */
+    // This function can reenter, but it doesn't pose a risk because _afterCall checks that the proposal is pending,
+    // thus any modifications to the operation during reentrancy should be caught.
+    // slither-disable-next-line reentrancy-eth
+    function execute(
+        address target,
+        uint256 value,
+        bytes calldata payload,
+        bytes32 predecessor,
+        bytes32 salt
+    ) public payable virtual onlyRoleOrOpenRole(EXECUTOR_ROLE) {
+        bytes32 id = hashOperation(target, value, payload, predecessor, salt);
+
+        _beforeCall(id, predecessor);
+        _execute(target, value, payload);
+        emit CallExecuted(id, 0, target, value, payload);
+        _afterCall(id);
+    }
+
+    /**
+     * @dev Execute an (ready) operation containing a batch of transactions.
+     *
+     * Emits one {CallExecuted} event per transaction in the batch.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'executor' role.
+     */
+    // This function can reenter, but it doesn't pose a risk because _afterCall checks that the proposal is pending,
+    // thus any modifications to the operation during reentrancy should be caught.
+    // slither-disable-next-line reentrancy-eth
+    function executeBatch(
+        address[] calldata targets,
+        uint256[] calldata values,
+        bytes[] calldata payloads,
+        bytes32 predecessor,
+        bytes32 salt
+    ) public payable virtual onlyRoleOrOpenRole(EXECUTOR_ROLE) {
+        require(
+            targets.length == values.length,
+            "TimelockController: length mismatch"
+        );
+        require(
+            targets.length == payloads.length,
+            "TimelockController: length mismatch"
+        );
+
+        bytes32 id = hashOperationBatch(
+            targets,
+            values,
+            payloads,
+            predecessor,
+            salt
+        );
+
+        _beforeCall(id, predecessor);
+        for (uint256 i = 0; i < targets.length; ++i) {
+            address target = targets[i];
+            uint256 value = values[i];
+            bytes calldata payload = payloads[i];
+            _execute(target, value, payload);
+            emit CallExecuted(id, i, target, value, payload);
+        }
+        _afterCall(id);
+    }
+
+    /**
+     * @dev Execute an operation's call.
+     */
+    function _execute(
+        address target,
+        uint256 value,
+        bytes calldata data
+    ) internal virtual {
+        (bool success, ) = target.call{value: value}(data);
+        require(success, "TimelockController: underlying transaction reverted");
+    }
+
+    /**
+     * @dev Checks before execution of an operation's calls.
+     */
+    function _beforeCall(bytes32 id, bytes32 predecessor) private view {
+        require(
+            isOperationReady(id),
+            "TimelockController: operation is not ready"
+        );
+        require(
+            predecessor == bytes32(0) || isOperationDone(predecessor),
+            "TimelockController: missing dependency"
+        );
+    }
+
+    /**
+     * @dev Checks after execution of an operation's calls.
+     */
+    function _afterCall(bytes32 id) private {
+        require(
+            isOperationReady(id),
+            "TimelockController: operation is not ready"
+        );
+        _timestamps[id] = _DONE_TIMESTAMP;
+    }
+
+    /**
+     * @dev Changes the minimum timelock duration for future operations.
+     *
+     * Emits a {MinDelayChange} event.
+     *
+     * Requirements:
+     *
+     * - the caller must be the timelock itself. This can only be achieved by scheduling and later executing
+     * an operation where the timelock is the target and the data is the ABI-encoded call to this function.
+     */
+    function updateDelay(uint256 newDelay) external virtual {
+        require(
+            msg.sender == address(this),
+            "TimelockController: caller must be timelock"
+        );
+        emit MinDelayChange(_minDelay, newDelay);
+        _minDelay = newDelay;
+    }
+
+    /**
+     * @dev See {IERC721Receiver-onERC721Received}.
+     */
+    function onERC721Received(
+        address,
+        address,
+        uint256,
+        bytes memory
+    ) public virtual override returns (bytes4) {
+        return this.onERC721Received.selector;
+    }
+
+    /**
+     * @dev See {IERC1155Receiver-onERC1155Received}.
+     */
+    function onERC1155Received(
+        address,
+        address,
+        uint256,
+        uint256,
+        bytes memory
+    ) public virtual override returns (bytes4) {
+        return this.onERC1155Received.selector;
+    }
+
+    /**
+     * @dev See {IERC1155Receiver-onERC1155BatchReceived}.
+     */
+    function onERC1155BatchReceived(
+        address,
+        address,
+        uint256[] memory,
+        uint256[] memory,
+        bytes memory
+    ) public virtual override returns (bytes4) {
+        return this.onERC1155BatchReceived.selector;
+    }
+}
diff --git a/hardhat.config.js b/hardhat.config.js
index c69737c..fcdb883 100644
--- a/hardhat.config.js
+++ b/hardhat.config.js
@@ -1,6 +1,7 @@
 /** @type import('hardhat/config').HardhatUserConfig */
 
 require('@nomicfoundation/hardhat-toolbox');
+require('@nomiclabs/hardhat-ethers');
 require('@openzeppelin/hardhat-upgrades');
 require('@nomiclabs/hardhat-solhint');
 require('dotenv').config();
diff --git a/scripts/100_migrateToIBTC.js b/scripts/100_migrateToIBTC.js
new file mode 100644
index 0000000..02c5f0c
--- /dev/null
+++ b/scripts/100_migrateToIBTC.js
@@ -0,0 +1,135 @@
+const hardhat = require('hardhat');
+const { ethers, upgrades } = require('hardhat');
+const {
+    loadContractAddress,
+    getExpectedContractAddress,
+    promptUser,
+    getMinimumDelay,
+} = require('./helpers/utils');
+const {
+    saveDeploymentInfo,
+    deploymentInfo,
+    loadDeploymentInfo,
+} = require('./helpers/deployment-handlers_versioned');
+const dlcAdminSafesConfigs = require('./helpers/dlc-admin-safes');
+const chalk = require('chalk');
+const safeContractProposal = require('./helpers/safe-api-service');
+
+const main = async () => {
+    const network = hardhat.network.name;
+    const accounts = await ethers.getSigners();
+    const deployer = accounts[0];
+    const dlcAdminSafes = dlcAdminSafesConfigs[network];
+    const proxyAdmin = await upgrades.admin.getInstance();
+    const proxyAdminOwner = await proxyAdmin.owner();
+    console.log('ProxyAdmin owner:', proxyAdminOwner);
+
+    const tokenProxyAddress = await loadContractAddress('DLCBTC', network);
+
+    const newImplementation = await ethers.getContractFactory('IBTC');
+
+    if (proxyAdminOwner == deployer.address) {
+        const ibtc = await upgrades.upgradeProxy(
+            tokenProxyAddress,
+            newImplementation
+        );
+        await ibtc.reinitializeEIP712();
+        await saveDeploymentInfo(deploymentInfo(network, ibtc, 'IBTC'));
+
+        await hardhat.run('verify:verify', {
+            address: ibtc.address,
+            constructorArguments: [],
+        });
+    } else {
+        console.log('New implementation:', newImplementation);
+        console.log(
+            'Expected contract address: ',
+            await getExpectedContractAddress(deployer)
+        );
+        if ((await promptUser('Are you sure you want to continue?')) === false)
+            return;
+        // We need to propose the upgrade through the SAFE & timelock
+        const newImplementationAddress = await upgrades.prepareUpgrade(
+            tokenProxyAddress,
+            newImplementation,
+            { timeout: 240 }
+        );
+        console.log('New implementation address', newImplementationAddress);
+
+        console.log('Verifying new implementation...');
+        await hardhat.run('verify:verify', {
+            address: newImplementationAddress,
+        });
+        console.log('New implementation verified.');
+
+        // we prepare the tx to the ProxyAdmin to upgrade the contract
+        // NOTE: we have to store this data for the actual execution
+        const upgradeTx = await proxyAdmin.populateTransaction.upgrade(
+            tokenProxyAddress,
+            newImplementationAddress
+        );
+        console.log('tokenProxyAddress', tokenProxyAddress);
+        console.log('newImplementationAddress', newImplementationAddress);
+        console.log('proxyAdmin.address', proxyAdmin.address);
+        console.log(
+            chalk.bgYellowBright('upgradeTx: (Store this!)'),
+            upgradeTx
+        );
+
+        // Fetching the TimelockController contract
+        const timeLockContractDeployInfo = await loadDeploymentInfo(
+            network,
+            'TimelockController'
+        );
+        const timelockContract = new hardhat.ethers.Contract(
+            timeLockContractDeployInfo.contract.address,
+            timeLockContractDeployInfo.contract.abi,
+            deployer
+        );
+
+        // Preparing the Multisig request to the TimelockController
+        const minimumDelay = getMinimumDelay(network);
+        const tlRequestParams = [
+            proxyAdmin.address,
+            0,
+            upgradeTx.data,
+            '0x0000000000000000000000000000000000000000000000000000000000000000',
+            '0x0000000000000000000000000000000000000000000000000000000000000000',
+            minimumDelay,
+        ];
+        const timelockContractTxRequest = await timelockContract
+            .connect(deployer)
+            .populateTransaction['schedule'](...tlRequestParams);
+        console.log('timelockContractTxRequest', timelockContractTxRequest);
+
+        // Proposing the upgrade through the SAFE
+        await safeContractProposal(
+            timelockContractTxRequest,
+            deployer,
+            dlcAdminSafes.critical
+        );
+
+        const implObject = await hardhat.ethers.getContractAt(
+            'IBTC',
+            // @ts-ignore
+            newImplementationAddress
+        );
+        const deploymentInfoToSave = deploymentInfo(
+            network,
+            { ...implObject, address: tokenProxyAddress },
+            'IBTC',
+            upgradeTx.data
+        );
+        await saveDeploymentInfo(
+            deploymentInfoToSave,
+            `deploymentFiles/${network}/IBTC.proposed.json`
+        );
+    }
+};
+
+main()
+    .then(() => process.exit(0))
+    .catch((error) => {
+        console.error(error);
+        process.exit(1);
+    });
diff --git a/scripts/50_contract-admin.js b/scripts/50_contract-admin.js
index fa55259..bbb4dfd 100644
--- a/scripts/50_contract-admin.js
+++ b/scripts/50_contract-admin.js
@@ -1,11 +1,17 @@
 require('dotenv').config();
 const hardhat = require('hardhat');
+const fs = require('fs/promises');
 
 const prompts = require('prompts');
 const chalk = require('chalk');
 const dlcAdminSafesConfigs = require('./helpers/dlc-admin-safes');
 const getContractConfigs = require('./99_contract-configs');
-const { promptUser, loadContractAddress } = require('./helpers/utils');
+const {
+    promptUser,
+    loadContractAddress,
+    getMinimumDelay,
+    getExpectedContractAddress,
+} = require('./helpers/utils');
 const {
     saveDeploymentInfo,
     deploymentInfo,
@@ -45,8 +51,23 @@ module.exports = async function contractAdmin() {
                 value: 'deploy',
             },
             { title: 'Verify Contract On Etherscan', value: 'verify' },
-            { title: 'Validate an Upgrade', value: 'validate-upgrade' },
-            { title: 'Upgrade Proxy Implementation', value: 'upgrade' },
+            {
+                title: 'Validate an Upgrade',
+                value: 'validate-upgrade',
+                description:
+                    'Validate a potential contract upgrade without performing any deployments.',
+            },
+            {
+                title: 'Upgrade Proxy',
+                value: 'upgrade',
+                description:
+                    'Upgrade a contract. Either directly, or through a SAFE proposal through a TimelockController',
+            },
+            {
+                title: 'Execute Upgrade',
+                value: 'execute-upgrade',
+                description: 'Execute an upgrade after the delay period',
+            },
             {
                 title: 'Transfer DLCBTC Ownership',
                 value: 'transfer-dlcbtc',
@@ -55,10 +76,13 @@ module.exports = async function contractAdmin() {
             {
                 title: 'Begin Transfer DEFAULT_ADMIN_ROLE',
                 value: 'transfer-admin',
+                description:
+                    'Begin the transfer of DEFAULT_ADMIN_ROLE on the DLCManager contract',
             },
             {
                 title: 'Transfer ProxyAdmin Ownership',
                 value: 'transfer-proxyadmin',
+                description: 'Transfer ownership of the ProxyAdmin contract',
             },
         ],
         initial: 0,
@@ -137,19 +161,15 @@ module.exports = async function contractAdmin() {
             }
             const contractSelectPrompt = await prompts({
                 type: 'select',
-                name: 'contracts',
+                name: 'contract',
                 message: `Select contract to verify on ${network}`,
                 choices: contractConfigs.map((config) => ({
                     title: `${config.name}`,
-                    value: config.name,
+                    value: config,
                 })),
             });
-            const contractName = contractSelectPrompt.contracts;
-            const contractConfig = contractConfigs.find(
-                (config) => config.name === contractName
-            );
             try {
-                await contractConfig.verify();
+                await contractSelectPrompt.contract.verify();
             } catch (error) {
                 console.error(error);
             }
@@ -176,15 +196,16 @@ module.exports = async function contractAdmin() {
             );
             const newImplementation =
                 await hardhat.ethers.getContractFactory(contractName);
-            const validation = await hardhat.upgrades.validateUpgrade(
-                proxyAddress,
-                newImplementation
-            );
-            if (!validation) {
-                console.log('Upgrade is valid');
+            try {
+                await hardhat.upgrades.validateUpgrade(
+                    proxyAddress,
+                    newImplementation
+                );
+            } catch (error) {
+                console.error(error);
                 return;
             }
-            console.log(validation);
+            console.log('Upgrade is valid');
             break;
         }
         case 'upgrade': {
@@ -199,32 +220,30 @@ module.exports = async function contractAdmin() {
                         value: config.name,
                     })),
             });
+            await hardhat.run('clean');
             await hardhat.run('compile');
             const contractName = contractSelectPrompt.contracts;
             const proxyAddress = await loadContractAddress(
                 contractName,
                 network
             );
-            const proxyAdminAddress =
-                await hardhat.upgrades.erc1967.getAdminAddress(proxyAddress);
 
-            const proxyAdmin = new hardhat.ethers.Contract(
-                proxyAdminAddress,
-                ['function owner() view returns (address)'],
-                deployer
-            );
+            const proxyAdmin = await hardhat.upgrades.admin.getInstance();
             const proxyAdminOwner = await proxyAdmin.owner();
             console.log('ProxyAdmin owner:', proxyAdminOwner);
 
             const newImplementation =
                 await hardhat.ethers.getContractFactory(contractName);
 
-            if (proxyAdminOwner == deployer) {
+            if (proxyAdminOwner == deployer.address) {
+                // Deployer can perform the whole upgrade flow
                 console.log('deployer is ProxyAdmin owner, continuing...');
                 await hardhat.upgrades.upgradeProxy(
                     proxyAddress,
                     newImplementation,
                     {
+                        timeout: 2000,
+                        // @ts-ignore
                         txOverrides: {
                             maxFeePerGas: 1000000000,
                             maxPriorityFeePerGas: 1000000000,
@@ -241,10 +260,37 @@ module.exports = async function contractAdmin() {
                     await saveDeploymentInfo(
                         deploymentInfo(network, contractObject, contractName)
                     );
+                    const contractConfig = contractConfigs.find(
+                        (config) => config.name === contractName
+                    );
+                    // @ts-ignore
+                    await contractConfig.verify();
+
+                    if (contractName == 'DLCManager') {
+                        if (
+                            (await promptUser(
+                                'Do you want to call initializeV2()?'
+                            )) === false
+                        )
+                            return;
+
+                        await contractObject.initializeV2();
+                    }
                 } catch (error) {
                     console.error(error);
                 }
             } else {
+                console.log('New implementation:', newImplementation);
+                console.log(
+                    'Expected contract address: ',
+                    await getExpectedContractAddress(deployer)
+                );
+                if (
+                    (await promptUser('Are you sure you want to continue?')) ===
+                    false
+                )
+                    return;
+                // We need to propose the upgrade through the SAFE & timelock
                 const newImplementationAddress =
                     await hardhat.upgrades.prepareUpgrade(
                         proxyAddress,
@@ -256,45 +302,163 @@ module.exports = async function contractAdmin() {
                     newImplementationAddress
                 );
 
-                try {
-                    await hardhat.upgrades.upgradeProxy(
-                        proxyAddress,
-                        newImplementation
-                    );
-                    console.log('Upgraded contract', contractName);
-                    console.log('Updating DeploymentInfo...');
-                    try {
-                        const contractObject =
-                            await hardhat.ethers.getContractAt(
-                                contractName,
-                                proxyAddress
-                            );
-                        await saveDeploymentInfo(
-                            deploymentInfo(
-                                network,
-                                contractObject,
-                                contractName
-                            )
-                        );
-                    } catch (error) {
-                        console.error(error);
-                    }
-                } catch (error) {
-                    console.error(error);
-                    console.log(chalk.bgYellow('Upgrade through the SAFE!'));
-                    const implObject = await hardhat.ethers.getContractAt(
-                        contractName,
-                        newImplementationAddress
-                    );
-                    const deploymentInfoToSave = deploymentInfo(
-                        network,
-                        { ...implObject, address: proxyAddress },
-                        contractName
-                    );
-                    await saveDeploymentInfo(deploymentInfoToSave);
-                    console.log(deploymentInfoToSave);
-                    console.log(chalk.bgRed('Upgrade through the SAFE!'));
-                }
+                console.log('Verifying new implementation...');
+                await hardhat.run('verify:verify', {
+                    address: newImplementationAddress,
+                });
+                console.log('New implementation verified.');
+
+                // we prepare the tx to the ProxyAdmin to upgrade the contract
+                // NOTE: we have to store this data for the actual execution
+                const upgradeTx = await proxyAdmin.populateTransaction.upgrade(
+                    proxyAddress,
+                    newImplementationAddress
+                );
+                console.log('proxyAddress', proxyAddress);
+                console.log(
+                    'newImplementationAddress',
+                    newImplementationAddress
+                );
+                console.log('proxyAdmin.address', proxyAdmin.address);
+                console.log(
+                    chalk.bgYellowBright('upgradeTx: (Store this!)'),
+                    upgradeTx
+                );
+
+                // Fetching the TimelockController contract
+                const timeLockContractDeployInfo = await loadDeploymentInfo(
+                    network,
+                    'TimelockController'
+                );
+                const timelockContract = new hardhat.ethers.Contract(
+                    timeLockContractDeployInfo.contract.address,
+                    timeLockContractDeployInfo.contract.abi,
+                    deployer
+                );
+
+                // Preparing the Multisig request to the TimelockController
+                const minimumDelay = getMinimumDelay(network);
+                const tlRequestParams = [
+                    proxyAdmin.address,
+                    0,
+                    upgradeTx.data,
+                    '0x0000000000000000000000000000000000000000000000000000000000000000',
+                    '0x0000000000000000000000000000000000000000000000000000000000000000',
+                    minimumDelay,
+                ];
+                const timelockContractTxRequest = await timelockContract
+                    .connect(deployer)
+                    .populateTransaction['schedule'](...tlRequestParams);
+                console.log(
+                    'timelockContractTxRequest',
+                    timelockContractTxRequest
+                );
+
+                // Proposing the upgrade through the SAFE
+                await safeContractProposal(
+                    timelockContractTxRequest,
+                    deployer,
+                    dlcAdminSafes.critical
+                );
+
+                const implObject = await hardhat.ethers.getContractAt(
+                    contractName,
+                    // @ts-ignore
+                    newImplementationAddress
+                );
+                const deploymentInfoToSave = deploymentInfo(
+                    network,
+                    { ...implObject, address: proxyAddress },
+                    contractName,
+                    upgradeTx.data
+                );
+                await saveDeploymentInfo(
+                    deploymentInfoToSave,
+                    `deploymentFiles/${network}/${contractName}.proposed.json`
+                );
+            }
+
+            break;
+        }
+        case 'execute-upgrade': {
+            const contractSelectPrompt = await prompts({
+                type: 'select',
+                name: 'contracts',
+                message: `Select contract to upgrade on ${network}`,
+                choices: contractConfigs
+                    .filter((config) => config.upgradeable)
+                    .map((config) => ({
+                        title: `${config.name}`,
+                        value: config.name,
+                    })),
+            });
+            const contractName = contractSelectPrompt.contracts;
+
+            const contractDeployInfo = await loadDeploymentInfo(
+                network,
+                contractName,
+                true
+            );
+
+            // Fetching the TimelockController contract
+            const timeLockContractDeployInfo = await loadDeploymentInfo(
+                network,
+                'TimelockController'
+            );
+            const timelockContract = new hardhat.ethers.Contract(
+                timeLockContractDeployInfo.contract.address,
+                timeLockContractDeployInfo.contract.abi,
+                deployer
+            );
+            const proxyAdmin = await hardhat.upgrades.admin.getInstance();
+            const tlRequestParams = [
+                proxyAdmin.address,
+                0,
+                contractDeployInfo.upgradeData,
+                '0x0000000000000000000000000000000000000000000000000000000000000000',
+                '0x0000000000000000000000000000000000000000000000000000000000000000',
+            ];
+
+            console.log('Executing upgrade...');
+            await timelockContract
+                .connect(deployer)
+                .execute(...tlRequestParams);
+
+            console.log('Updating DeploymentInfo...');
+            await fs.copyFile(
+                `deploymentFiles/${network}/${contractName}.proposed.json`,
+                `deploymentFiles/${network}/${contractName}.json`
+            );
+            await fs.rename(
+                `deploymentFiles/${network}/${contractName}.proposed.json`,
+                `deploymentFiles/${network}/${contractName}.${new Date().toISOString()}.json`
+            );
+            console.log('DeploymentInfo updated.');
+
+            if (contractName == 'IBTC') {
+                if (
+                    (await promptUser(
+                        'Do you want to call reinitializeEIP712 on iBTC?'
+                    )) === false
+                )
+                    return;
+                const iBTC = await hardhat.ethers.getContractAt(
+                    'IBTC',
+                    contractDeployInfo.contract.address
+                );
+                await iBTC.reinitializeEIP712();
+            } else if (contractName == 'DLCManager') {
+                if (
+                    (await promptUser(
+                        'Do you want to call initializeV2()?'
+                    )) === false
+                )
+                    return;
+                const dlcManager = await hardhat.ethers.getContractAt(
+                    'DLCManager',
+                    contractDeployInfo.contract.address
+                );
+                await dlcManager.initializeV2();
             }
 
             break;
@@ -386,21 +550,55 @@ module.exports = async function contractAdmin() {
             break;
         }
         case 'transfer-proxyadmin': {
-            const currentAdmin = await (
-                await hardhat.upgrades.admin.getInstance()
-            ).functions['owner()']();
+            const proxyAdmin = await hardhat.upgrades.admin.getInstance();
+            const currentAdminOwner = await proxyAdmin.functions['owner()']();
 
             console.log(
-                chalk.bgYellow('Current ProxyAdmin owner:', currentAdmin)
+                chalk.bgYellow('Current ProxyAdmin owner:', currentAdminOwner)
             );
-            if (currentAdmin == dlcAdminSafes.critical) return;
+            if (currentAdminOwner == dlcAdminSafes.critical) {
+                console.log(
+                    chalk.bgRed(
+                        'Current ProxyAdmin owner is the Critical Multisig Already!'
+                    )
+                );
+                if (
+                    (await promptUser('Are you sure you want to continue?')) ===
+                    false
+                )
+                    return;
+            }
 
             const newAdmin = await prompts({
                 type: 'text',
                 name: 'value',
-                message: 'Enter new ProxyAdmin address',
+                message:
+                    'Enter new ProxyAdmin address or use default TimelockController address',
+                initial: await loadContractAddress(
+                    'TimelockController',
+                    network
+                ),
             });
             if (!newAdmin.value) return;
+
+            if (currentAdminOwner != deployer.address) {
+                const txRequest =
+                    await proxyAdmin.populateTransaction.transferOwnership(
+                        newAdmin.value
+                    );
+                console.log(
+                    'Proposing ownership transfer of ProxyAdmin to:',
+                    newAdmin.value
+                );
+                console.log('txRequest', txRequest);
+                await safeContractProposal(
+                    txRequest,
+                    deployer,
+                    dlcAdminSafes.critical
+                );
+                return;
+            }
+
             console.log('Transferring ownership of ProxyAdmin...');
             await hardhat.upgrades.admin.transferProxyAdminOwnership(
                 newAdmin.value
diff --git a/scripts/99_contract-configs.js b/scripts/99_contract-configs.js
index fe0e0ab..af2ccf9 100644
--- a/scripts/99_contract-configs.js
+++ b/scripts/99_contract-configs.js
@@ -4,7 +4,11 @@ const {
     saveDeploymentInfo,
     deploymentInfo,
 } = require('./helpers/deployment-handlers_versioned');
-const { promptUser, loadContractAddress } = require('./helpers/utils');
+const {
+    promptUser,
+    loadContractAddress,
+    getMinimumDelay,
+} = require('./helpers/utils');
 
 // This is a pure function that just logs
 async function beforeDeployment(contractName, constructorArguments, network) {
@@ -37,32 +41,77 @@ module.exports = function getContractConfigs(networkConfig, _btcFeeRecipient) {
     const btcFeeRecipient =
         _btcFeeRecipient ?? '0014e60f61fa2f2941217934d5f9976bf27381b3b036';
     const threshold = 2;
+    const minimumDelay = getMinimumDelay(networkName);
+    const proposers = [dlcAdminSafes.critical];
+    const executors = [dlcAdminSafes.critical, deployer.address];
+    const timelockConstructorArgs = [
+        minimumDelay,
+        proposers,
+        executors,
+        hardhat.ethers.constants.AddressZero, // no admin
+    ];
 
     return [
         {
-            name: 'DLCBTC',
+            name: 'IBTC',
             deployer: deployer.address,
             upgradeable: true,
             requirements: [],
             deploy: async (requirementAddresses) => {
-                await beforeDeployment('DLCBTC', '', networkName);
+                await beforeDeployment('IBTC', '', networkName);
+
+                const IBTC = await hardhat.ethers.getContractFactory('IBTC');
+                const iBTC = await hardhat.upgrades.deployProxy(IBTC);
+                await iBTC.deployed();
+
+                await afterDeployment('IBTC', iBTC, networkName);
+
+                return iBTC.address;
+            },
+            verify: async () => {
+                const address = await loadContractAddress('IBTC', networkName);
+                await hardhat.run('verify:verify', {
+                    address: address,
+                });
+            },
+        },
+        {
+            name: 'TimelockController',
+            deployer: deployer.address,
+            upgradeable: false,
+            requirements: [],
+            deploy: async (requirementAddresses) => {
+                await beforeDeployment(
+                    'TimelockController',
+                    timelockConstructorArgs,
+                    networkName
+                );
 
-                const DLCBTC =
-                    await hardhat.ethers.getContractFactory('DLCBTC');
-                const dlcBtc = await hardhat.upgrades.deployProxy(DLCBTC);
-                await dlcBtc.deployed();
+                const TimelockController =
+                    await hardhat.ethers.getContractFactory(
+                        'TimelockController'
+                    );
+                const timelockController = await TimelockController.deploy(
+                    ...timelockConstructorArgs
+                );
+                await timelockController.deployed();
 
-                await afterDeployment('DLCBTC', dlcBtc, networkName);
+                await afterDeployment(
+                    'TimelockController',
+                    timelockController,
+                    networkName
+                );
 
-                return dlcBtc.address;
+                return timelockController.address;
             },
             verify: async () => {
                 const address = await loadContractAddress(
-                    'DLCBTC',
+                    'TimelockController',
                     networkName
                 );
                 await hardhat.run('verify:verify', {
                     address: address,
+                    constructorArguments: timelockConstructorArgs,
                 });
             },
         },
@@ -70,21 +119,20 @@ module.exports = function getContractConfigs(networkConfig, _btcFeeRecipient) {
             name: 'DLCManager',
             deployer: deployer.address,
             upgradeable: true,
-            requirements: ['DLCBTC'],
+            requirements: ['IBTC'],
             deploy: async (requirementAddresses) => {
                 // const defaultAdmin = deployer.address;
                 // const dlcAdmin = deployer.address;
                 const defaultAdmin = dlcAdminSafes.critical;
                 const dlcAdmin = dlcAdminSafes.medium;
-                const DLCBTCAddress = requirementAddresses['DLCBTC'];
-                if (!DLCBTCAddress)
-                    throw new Error('DLCBTC deployment not found.');
+                const IBTCAddress = requirementAddresses['IBTC'];
+                if (!IBTCAddress) throw new Error('IBTC deployment not found.');
                 await beforeDeployment(
                     'DLCManager',
                     `defaultAdmin: ${defaultAdmin}, \
                     dlcAdminRole: ${dlcAdmin}, \
                     threshold: ${threshold}, \
-                    tokenContract: ${DLCBTCAddress}, \
+                    tokenContract: ${IBTCAddress}, \
                     btcFeeRecipient: ${btcFeeRecipient}`,
                     networkName
                 );
@@ -96,7 +144,7 @@ module.exports = function getContractConfigs(networkConfig, _btcFeeRecipient) {
                         defaultAdmin,
                         dlcAdmin,
                         threshold,
-                        DLCBTCAddress,
+                        IBTCAddress,
                         btcFeeRecipient,
                     ]
                 );
@@ -104,29 +152,29 @@ module.exports = function getContractConfigs(networkConfig, _btcFeeRecipient) {
 
                 await afterDeployment('DLCManager', dlcManager, networkName);
 
-                const dlcBtc = await hardhat.ethers.getContractAt(
-                    'DLCBTC',
-                    DLCBTCAddress
+                const iBTC = await hardhat.ethers.getContractAt(
+                    'IBTC',
+                    IBTCAddress
                 );
-                const currentOwner = await dlcBtc.owner();
+                const currentOwner = await iBTC.owner();
                 console.log(
-                    chalk.bgYellow('Current DLCBTC owner:', currentOwner)
+                    chalk.bgYellow('Current IBTC owner:', currentOwner)
                 );
 
                 if (currentOwner === dlcManager.address) {
                     console.log(
-                        'DLCBTC is already owned by DLCManager, skipping transfer...'
+                        'IBTC is already owned by DLCManager, skipping transfer...'
                     );
                 } else {
                     const shouldTransferOwnership = await promptUser(
-                        `Would you like to transfer ownership of DLCBTC contract to ${dlcManager.address}?`
+                        `Would you like to transfer ownership of IBTC contract to ${dlcManager.address}?`
                     );
                     if (shouldTransferOwnership) {
                         if (currentOwner === deployer.address) {
                             console.log(
-                                'DLCBTC is owned by deployer, transferring ownership...'
+                                'IBTC is owned by deployer, transferring ownership...'
                             );
-                            const tx = await dlcBtc
+                            const tx = await iBTC
                                 .connect(deployer)
                                 .transferOwnership(dlcManager.address);
                             await tx.wait();
@@ -145,8 +193,8 @@ module.exports = function getContractConfigs(networkConfig, _btcFeeRecipient) {
                             console.log(receipt);
                         }
 
-                        const newOwner = await dlcBtc.owner();
-                        console.log('New DLCBTC Owner:', newOwner);
+                        const newOwner = await iBTC.owner();
+                        console.log('New IBTC Owner:', newOwner);
                     }
                 }
 
@@ -154,7 +202,7 @@ module.exports = function getContractConfigs(networkConfig, _btcFeeRecipient) {
             },
             verify: async () => {
                 const address = await loadContractAddress(
-                    'DlcManager',
+                    'DLCManager',
                     networkName
                 );
                 await hardhat.run('verify:verify', {
diff --git a/scripts/helpers/chainlink-por-addresses.js b/scripts/helpers/chainlink-por-addresses.js
index 814d0dc..d4eb8de 100644
--- a/scripts/helpers/chainlink-por-addresses.js
+++ b/scripts/helpers/chainlink-por-addresses.js
@@ -18,7 +18,7 @@ module.exports = async function getPoRAddress(network) {
         case 'arbitrum':
             return '0x47A2fBEb46553F01E7133686Fb1b5349d4823a0C';
         case 'arbsepolia':
-            return '0xeA7cE82EC8568a19E8a92b9E588f52a1bEC96F3D';
+            return '';
         case 'base':
             return '';
         case 'basesepolia':
diff --git a/scripts/helpers/deployment-handlers_versioned.js b/scripts/helpers/deployment-handlers_versioned.js
index bb45f42..ad1eddd 100644
--- a/scripts/helpers/deployment-handlers_versioned.js
+++ b/scripts/helpers/deployment-handlers_versioned.js
@@ -1,7 +1,7 @@
 const fs = require('fs/promises');
 const { execSync } = require('child_process');
 
-function deploymentInfo(networkName, contract, contractName) {
+function deploymentInfo(networkName, contract, contractName, upgradeData) {
     let _gitSHA;
     try {
         _gitSHA = execSync('git rev-parse --short HEAD').toString().trim();
@@ -19,12 +19,13 @@ function deploymentInfo(networkName, contract, contractName) {
             signerAddress: contract.signer.address,
             abi: contract.interface.format(),
         },
+        upgradeData: upgradeData || '',
     };
     console.log(deployInfo);
     return deployInfo;
 }
 
-async function saveDeploymentInfo(info, filename = undefined) {
+async function saveDeploymentInfo(info, filename) {
     if (!filename) {
         filename = `deploymentFiles/${info.network}/${info.contract.name}.json`;
     }
@@ -56,8 +57,10 @@ function validateDeploymentInfo(deployInfo) {
     required('abi');
 }
 
-async function loadDeploymentInfo(networkName, contractName) {
-    const deploymentConfigFile = `deploymentFiles/${networkName}/${contractName}.json`;
+async function loadDeploymentInfo(networkName, contractName, proposed) {
+    const deploymentConfigFile = `deploymentFiles/${networkName}/${contractName}${
+        proposed ? '.proposed' : ''
+    }.json`;
     const content = await fs.readFile(deploymentConfigFile, {
         encoding: 'utf8',
     });
diff --git a/scripts/helpers/dlc-admin-safes.js b/scripts/helpers/dlc-admin-safes.js
index 1672b0f..a9ab7ee 100644
--- a/scripts/helpers/dlc-admin-safes.js
+++ b/scripts/helpers/dlc-admin-safes.js
@@ -29,6 +29,6 @@ module.exports = {
     },
     basesepolia: {
         medium: '0x9506Ea24038609679732855F757041a3C1C06623',
-        critical: '0x9506Ea24038609679732855F757041a3C1C06623',
+        critical: '0xebDC2027D3ee493B49553Befc1200e1cce9e2E08',
     },
 };
diff --git a/scripts/helpers/mainnet-call-all.sh b/scripts/helpers/mainnet-call-all.sh
index 431b6e9..045a352 100755
--- a/scripts/helpers/mainnet-call-all.sh
+++ b/scripts/helpers/mainnet-call-all.sh
@@ -13,3 +13,4 @@ extra_args="${@:2}"
 HARDHAT_NETWORK=base dlc-link-eth $command $extra_args
 HARDHAT_NETWORK=arbitrum dlc-link-eth $command $extra_args
 HARDHAT_NETWORK=mainnet dlc-link-eth $command $extra_args
+HARDHAT_NETWORK=optimism dlc-link-eth $command $extra_args
diff --git a/scripts/helpers/new-chain-setup-testnet.sh b/scripts/helpers/new-chain-setup-testnet.sh
deleted file mode 100755
index ba2a564..0000000
--- a/scripts/helpers/new-chain-setup-testnet.sh
+++ /dev/null
@@ -1,25 +0,0 @@
-#!/bin/bash
-
-# Set CLI_MODE='noninteractive' in the environment
-export CLI_MODE='noninteractive'
-
-# Testnet Attestors
-dlc-link-eth add-signer 0x3355977947F84C2b1CAE7D2903a72958aEE185e2
-dlc-link-eth add-signer 0xA2c975F49f578AC3E384Ad21E2126968b82037f5
-dlc-link-eth add-signer 0x3fF31dCd8ca84b594e355CfDf913738bFc192a5c
-dlc-link-eth add-signer 0xc5E34272789180802cd7610Ef203EB196CB3522D # Stakin
-dlc-link-eth add-signer 0xAE9749CE4616193c7Bf8A1A395Db3C7f910120d6 # LinkPool
-dlc-link-eth add-signer 0xABD434AA01eE1A6d3E4a569B871E7e63E64F166c # HashKey
-dlc-link-eth add-signer 0x6a79E12566E01e8Fda80a15b7325478B1d68B342 # Despread
-dlc-link-eth add-signer 0xB626fa1355B4c64Ca9723D555Bd1AA9eD8491cB4 # ValidationCloud
-
-dlc-link-eth set-whitelisting 'false'
-
-dlc-link-eth set-attestor-gpk 'tpubDDRekL64eJJav32TLhNhG59qra7wAMaei8YMGXNiJE8ksdYrKgvaFM1XG6JrSt31W97XryScrX37RUEujjZT4qScNf8Zu1JxWj4VYkwz4rU'
-
-dlc-link-eth set-btc-fee-recipient 032392b61a5c3b0098774465ad61e429fd892615ff2890f849f8eb237a8a59f3ba
-
-dlc-link-eth set-threshold 4
-
-# Vault setup and print dlcUUID
-# dlc-link-eth setup-vault 1200000 | grep dlcUUID
diff --git a/scripts/helpers/safe-api-service.js b/scripts/helpers/safe-api-service.js
index d1dd074..00cea1a 100644
--- a/scripts/helpers/safe-api-service.js
+++ b/scripts/helpers/safe-api-service.js
@@ -12,7 +12,10 @@ module.exports = async function safeContractProposal(
     signer,
     safeAddress
 ) {
-    const network = hardhat.network.name;
+    let network = hardhat.network.name;
+
+    if (network === 'basesepolia') network = 'base-sepolia';
+
     console.log('Network', network);
 
     const ethAdapter = new EthersAdapter({
diff --git a/scripts/helpers/utils.js b/scripts/helpers/utils.js
index 5947327..2d33328 100644
--- a/scripts/helpers/utils.js
+++ b/scripts/helpers/utils.js
@@ -1,6 +1,7 @@
 require('dotenv').config();
 const prompts = require('prompts');
 const { loadDeploymentInfo } = require('./deployment-handlers_versioned');
+const { ethers } = require('hardhat');
 
 async function promptUser(message) {
     if (process.env.CLI_MODE === 'noninteractive') {
@@ -29,7 +30,39 @@ async function loadContractAddress(requirement, network) {
     return deployment.contract.address;
 }
 
+function getMinimumDelay(networkName) {
+    switch (networkName) {
+        case 'localhost':
+        case 'arbsepolia':
+        case 'sepolia':
+        case 'basesepolia':
+            return 60 * 2; // 2 minutes
+        default:
+            return 60 * 60 * 24 * 7; // 1 week
+    }
+}
+
+async function getExpectedContractAddress(deployer, nonce = undefined) {
+    // If nonce is not provided, get the current nonce
+    if (nonce === undefined) {
+        nonce = await deployer.getTransactionCount();
+    }
+
+    // Calculate the address using CREATE1 formula
+    const addressBytes = ethers.utils.keccak256(
+        ethers.utils.RLP.encode([
+            deployer.address,
+            ethers.BigNumber.from(nonce).toHexString(),
+        ])
+    );
+
+    // Take the last 40 characters (20 bytes) of the hash
+    return ethers.utils.getAddress('0x' + addressBytes.substring(26));
+}
+
 module.exports = {
     promptUser,
     loadContractAddress,
+    getMinimumDelay,
+    getExpectedContractAddress,
 };
diff --git a/test/DLCBTC.test.js b/test/DLCBTC.test.js
index 564e628..da3a725 100644
--- a/test/DLCBTC.test.js
+++ b/test/DLCBTC.test.js
@@ -10,8 +10,8 @@ const mockBTCTxId =
 const mockTaprootPubkey =
     '0x1234567890123456789012345678901234567890123456789012345678901234';
 
-describe('DLCBTC', function () {
-    let dlcBtc, dlcManager;
+describe('iBTC', function () {
+    let iBTC, dlcManager;
     let accounts, deployer, user, someRandomAccount;
     let attestor1, attestor2, attestor3;
     let attestors;
@@ -30,9 +30,9 @@ describe('DLCBTC', function () {
         attestor3 = accounts[8];
         attestors = [attestor1, attestor2, attestor3];
 
-        const DLCBTC = await ethers.getContractFactory('DLCBTC', deployer);
-        dlcBtc = await upgrades.deployProxy(DLCBTC);
-        await dlcBtc.deployed();
+        const IBTC = await ethers.getContractFactory('IBTC', deployer);
+        iBTC = await upgrades.deployProxy(IBTC);
+        await iBTC.deployed();
 
         // DLCManager
         const DLCManager = await ethers.getContractFactory('DLCManager');
@@ -40,79 +40,79 @@ describe('DLCBTC', function () {
             deployer.address,
             deployer.address,
             3,
-            dlcBtc.address,
+            iBTC.address,
             btcFeeRecipient,
         ]);
         await dlcManager.deployed();
     });
 
     it('should deploy', async () => {
-        expect(dlcBtc.address).to.not.equal(0);
+        expect(iBTC.address).to.not.equal(0);
     });
 
     it('should be owned by deployer at start', async () => {
-        expect(await dlcBtc.owner()).to.equal(deployer.address);
+        expect(await iBTC.owner()).to.equal(deployer.address);
     });
 
     it('should have 8 decimals', async () => {
-        expect(await dlcBtc.decimals()).to.equal(8);
+        expect(await iBTC.decimals()).to.equal(8);
     });
 
     it('should have 0 total supply', async () => {
-        expect(await dlcBtc.totalSupply()).to.equal(0);
+        expect(await iBTC.totalSupply()).to.equal(0);
     });
 
     it('should revert on unauthorized mint', async () => {
         await expect(
-            dlcBtc.connect(user)['mint(address,uint256)'](user.address, deposit)
-        ).to.be.revertedWithCustomError(dlcBtc, 'NotAuthorized');
+            iBTC.connect(user)['mint(address,uint256)'](user.address, deposit)
+        ).to.be.revertedWithCustomError(iBTC, 'NotAuthorized');
     });
 
     it('should revert on unauthorized burn', async () => {
         await expect(
-            dlcBtc.connect(user)['burn(address,uint256)'](user.address, deposit)
+            iBTC.connect(user)['burn(address,uint256)'](user.address, deposit)
         ).to.be.revertedWith('Ownable: caller is not the owner');
     });
 
     it('owner can mint tokens', async () => {
-        await dlcBtc['mint(address,uint256)'](user.address, deposit);
-        expect(await dlcBtc.balanceOf(user.address)).to.equal(deposit);
+        await iBTC['mint(address,uint256)'](user.address, deposit);
+        expect(await iBTC.balanceOf(user.address)).to.equal(deposit);
     });
 
     it('owner can burn tokens', async () => {
-        await dlcBtc['mint(address,uint256)'](user.address, deposit);
-        await dlcBtc['burn(address,uint256)'](user.address, deposit);
-        expect(await dlcBtc.balanceOf(user.address)).to.equal(0);
+        await iBTC['mint(address,uint256)'](user.address, deposit);
+        await iBTC['burn(address,uint256)'](user.address, deposit);
+        expect(await iBTC.balanceOf(user.address)).to.equal(0);
     });
 
     describe('after Ownership transfer', async () => {
         beforeEach(async () => {
-            await dlcBtc['mint(address,uint256)'](user.address, deposit);
-            await dlcBtc.transferOwnership(dlcManager.address);
+            await iBTC['mint(address,uint256)'](user.address, deposit);
+            await iBTC.transferOwnership(dlcManager.address);
         });
 
         it('should be owned by dlcManager', async () => {
-            expect(await dlcBtc.owner()).to.equal(dlcManager.address);
+            expect(await iBTC.owner()).to.equal(dlcManager.address);
         });
 
         it('should revert on mint called by previous owner', async () => {
             await expect(
-                dlcBtc
+                iBTC
                     .connect(deployer)
                     ['mint(address,uint256)'](user.address, deposit)
-            ).to.be.revertedWithCustomError(dlcBtc, 'NotAuthorized');
+            ).to.be.revertedWithCustomError(iBTC, 'NotAuthorized');
         });
 
         it('should revert on burn called by previous owner', async () => {
             await expect(
-                dlcBtc
+                iBTC
                     .connect(deployer)
                     ['burn(address,uint256)'](user.address, deposit)
             ).to.be.revertedWith('Ownable: caller is not the owner');
         });
 
         it('dlcManager can mint tokens', async () => {
-            const existingBalance = await dlcBtc.balanceOf(user.address);
+            const existingBalance = await iBTC.balanceOf(user.address);
             await dlcManager.connect(deployer).whitelistAddress(user.address);
             const tx = await dlcManager.connect(user).setupVault();
             const receipt = await tx.wait();
@@ -161,13 +161,13 @@ describe('DLCBTC', function () {
             const expectedBalance = ethers.BigNumber.from(existingBalance).add(
                 ethers.BigNumber.from(deposit)
             );
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(
+            expect(await iBTC.balanceOf(user.address)).to.equal(
                 expectedBalance
             );
         });
 
         it('dlcManager can burn tokens', async () => {
-            const existingBalance = await dlcBtc.balanceOf(user.address);
+            const existingBalance = await iBTC.balanceOf(user.address);
             await dlcManager.connect(deployer).whitelistAddress(user.address);
             const tx = await dlcManager.connect(user).setupVault();
             const receipt = await tx.wait();
@@ -213,7 +213,7 @@ describe('DLCBTC', function () {
                     deposit
                 );
             await tx3.wait();
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(
+            expect(await iBTC.balanceOf(user.address)).to.equal(
                 ethers.BigNumber.from(existingBalance).add(
                     ethers.BigNumber.from(deposit)
                 )
@@ -223,7 +223,7 @@ describe('DLCBTC', function () {
                 .withdraw(mockUUID, deposit);
             await tx4.wait();
 
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(
+            expect(await iBTC.balanceOf(user.address)).to.equal(
                 existingBalance
             );
         });
diff --git a/test/DLCManager.test.js b/test/DLCManager.test.js
index b10b8ab..f735bcf 100644
--- a/test/DLCManager.test.js
+++ b/test/DLCManager.test.js
@@ -14,7 +14,7 @@ async function whitelistAddress(dlcManager, user) {
 }
 
 describe('DLCManager', () => {
-    let dlcManager, dlcBtc, uuid;
+    let dlcManager, iBTC, uuid;
     let accounts, deployer, user, randomAccount, anotherAccount, protocol;
     let attestor1, attestor2, attestor3;
     let attestors;
@@ -46,9 +46,9 @@ describe('DLCManager', () => {
         attestor3 = accounts[8];
         attestors = [attestor1, attestor2, attestor3];
 
-        const DLCBTC = await ethers.getContractFactory('DLCBTC', deployer);
-        dlcBtc = await hardhat.upgrades.deployProxy(DLCBTC);
-        await dlcBtc.deployed();
+        const IBTC = await ethers.getContractFactory('IBTC', deployer);
+        iBTC = await hardhat.upgrades.deployProxy(IBTC);
+        await iBTC.deployed();
 
         // DLCManager
         const DLCManager = await ethers.getContractFactory('DLCManager');
@@ -56,12 +56,12 @@ describe('DLCManager', () => {
             deployer.address,
             deployer.address,
             3,
-            dlcBtc.address,
+            iBTC.address,
             btcFeeRecipient,
         ]);
         await dlcManager.deployed();
 
-        await dlcBtc.transferOwnership(dlcManager.address);
+        await iBTC.transferOwnership(dlcManager.address);
     });
 
     describe('test contracts are deployed correctly', async () => {
@@ -69,8 +69,8 @@ describe('DLCManager', () => {
             expect(dlcManager.address).to.not.equal(0);
         });
 
-        it('should be the owner of the dlcBTC token contract', async () => {
-            expect(await dlcBtc.owner()).to.equal(dlcManager.address);
+        it('should be the owner of the iBTC token contract', async () => {
+            expect(await iBTC.owner()).to.equal(dlcManager.address);
         });
     });
 
@@ -499,7 +499,7 @@ describe('DLCManager', () => {
         });
 
         it('should revert on nonce-manipulated signatures from the same signer', async () => {
-            const existingBalance = await dlcBtc.balanceOf(user.address);
+            const existingBalance = await iBTC.balanceOf(user.address);
             const deposit = 100000000; // 1 BTC
             const tx = await dlcManager.connect(user).setupVault();
             const receipt = await tx.wait();
@@ -687,7 +687,7 @@ describe('DLCManager', () => {
             ).to.be.revertedWithCustomError(dlcManager, 'InvalidSigner');
         });
 
-        it('mints dlcBTC tokens to the user', async () => {
+        it('mints iBTC tokens to the user', async () => {
             await whitelistAddress(dlcManager, user);
             const tx = await dlcManager.connect(user).setupVault();
             await tx.wait();
@@ -706,7 +706,7 @@ describe('DLCManager', () => {
                 .connect(attestor1)
                 .setStatusFunded(uuid, btcTxId, signatureBytes, valueLocked);
             await tx2.wait();
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(valueLocked);
+            expect(await iBTC.balanceOf(user.address)).to.equal(valueLocked);
         });
 
         it('emits a StatusFunded event with the correct data', async () => {
@@ -734,7 +734,7 @@ describe('DLCManager', () => {
         });
     });
 
-    describe('getTotalValueMintedInVaults', async () => {
+    describe('totalValueMinted', async () => {
         beforeEach(async () => {
             await whitelistAddress(dlcManager, user);
 
@@ -788,8 +788,7 @@ describe('DLCManager', () => {
         });
 
         it('returns the correct value for 1 vault', async () => {
-            const totalValueMinted =
-                await dlcManager.getTotalValueMintedInVaults();
+            const totalValueMinted = await dlcManager.totalValueMinted();
             expect(totalValueMinted).to.equal(valueLocked);
         });
 
@@ -843,10 +842,19 @@ describe('DLCManager', () => {
                 );
             await tx3.wait();
 
-            const totalValueMinted =
-                await dlcManager.getTotalValueMintedInVaults();
+            const totalValueMinted = await dlcManager.totalValueMinted();
             expect(totalValueMinted).to.equal(valueLocked * 2);
         });
+
+        it('returns correct value after a withdrawal', async () => {
+            const tx = await dlcManager
+                .connect(user)
+                .withdraw(uuid, valueLocked);
+            await tx.wait();
+
+            const totalValueMinted = await dlcManager.totalValueMinted();
+            expect(totalValueMinted).to.equal(0);
+        });
     });
 
     describe('Proof of Reserves', async () => {
@@ -970,7 +978,7 @@ describe('DLCManager', () => {
                     signatureBytesForFunding,
                     valueLocked
                 );
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(valueLocked);
+            expect(await iBTC.balanceOf(user.address)).to.equal(valueLocked);
         });
     });
 
@@ -1034,7 +1042,7 @@ describe('DLCManager', () => {
                 .withdraw(uuid, valueLocked / 2);
             const getDlcTx = await dlcManager.getDLC(uuid);
 
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(
+            expect(await iBTC.balanceOf(user.address)).to.equal(
                 valueLocked / 2
             );
             expect(getDlcTx.valueMinted).to.equal(valueLocked / 2);
@@ -1089,7 +1097,7 @@ describe('DLCManager', () => {
                 );
             await tx3.wait();
 
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(
+            expect(await iBTC.balanceOf(user.address)).to.equal(
                 valueLocked / 2
             );
 
@@ -1292,7 +1300,7 @@ describe('DLCManager', () => {
                 );
             await tx3.wait();
 
-            expect(await dlcBtc.balanceOf(user.address)).to.equal(
+            expect(await iBTC.balanceOf(user.address)).to.equal(
                 lockedAmountAfterDeposit
             );
 
